File: by-iara-web/apps/admin/src/app/finance/finance.models.ts
export type ExpenseCategory =
  | 'RENT_UTILITIES'
  | 'SUPPLIES'
  | 'SOFTWARE'
  | 'MARKETING'
  | 'PAYMENT_FEES'
  | 'INSURANCE_LICENSES'
  | 'TRAVEL'
  | 'CONTRACTORS'
  | 'OTHER';

export type ExpenseStatus = 'ACTIVE' | 'VOIDED';
export type TrendGranularity = 'DAILY' | 'MONTHLY';

export interface Expense {
  id: string;
  category: ExpenseCategory;
  amountCents: number;
  currency: string;
  incurredAt: string;
  vendor: string | null;
  description: string;
  status: ExpenseStatus;
  voidedAt: string | null;
  createdAt: string;
}

export interface ExpenseInput {
  category: ExpenseCategory;
  amountCents: number;
  currency: string;
  incurredAt: string;
  vendor?: string;
  description: string;
}

export interface ExpensePage {
  items: Expense[];
  page: number;
  size: number;
  total: number;
}

export interface PaymentMethodTotal {
  method: 'CASH' | 'CARD' | 'BANK_TRANSFER' | 'OTHER';
  amountCents: number;
}

export interface IncomePayment {
  id: string;
  reservationId: string;
  customerName: string;
  serviceName: string;
  amountCents: number;
  currency: string;
  method: PaymentMethodTotal['method'];
  paidAt: string;
}

export interface IncomePaymentPage {
  items: IncomePayment[];
  page: number;
  size: number;
  total: number;
}

export interface FinancialTrendPoint {
  periodStart: string;
  revenueCents: number;
  expenseCents: number;
  profitCents: number;
}

export interface FinancialReport {
  from: string;
  to: string;
  currency: string;
  revenueCents: number;
  expenseCents: number;
  operatingProfitCents: number;
  outstandingBalanceCents: number;
  completedAppointments: number;
  noShows: number;
  averageCompletedValueCents: number;
  granularity: TrendGranularity;
  revenueByPaymentMethod: PaymentMethodTotal[];
  trend: FinancialTrendPoint[];
}

export const expenseCategoryLabels: Record<ExpenseCategory, string> = {
  RENT_UTILITIES: 'Rent & utilities',
  SUPPLIES: 'Supplies & products',
  SOFTWARE: 'Software & subscriptions',
  MARKETING: 'Marketing',
  PAYMENT_FEES: 'Payment fees',
  INSURANCE_LICENSES: 'Insurance & licences',
  TRAVEL: 'Travel',
  CONTRACTORS: 'Contractors & staff',
  OTHER: 'Other',
};

File: by-iara-api/src/main/kotlin/com/byiara/api/finance/domain/FinanceModels.kt
package com.byiara.api.finance.domain

import com.byiara.api.reservation.domain.PaymentMethod
import java.time.LocalDate
import java.time.OffsetDateTime
import java.util.UUID

enum class ExpenseCategory {
    RENT_UTILITIES,
    SUPPLIES,
    SOFTWARE,
    MARKETING,
    PAYMENT_FEES,
    INSURANCE_LICENSES,
    TRAVEL,
    CONTRACTORS,
    OTHER,
}

enum class ExpenseStatus {
    ACTIVE,
    VOIDED,
}

data class Expense(
    val id: UUID,
    val category: ExpenseCategory,
    val amountCents: Long,
    val currency: String,
    val incurredAt: OffsetDateTime,
    val vendor: String?,
    val description: String,
    val status: ExpenseStatus,
    val voidedAt: OffsetDateTime?,
    val createdAt: OffsetDateTime,
)

data class NewExpense(
    val category: ExpenseCategory,
    val amountCents: Long,
    val currency: String,
    val incurredAt: OffsetDateTime,
    val vendor: String?,
    val description: String,
)

enum class TrendGranularity {
    DAILY,
    MONTHLY,
}

data class PaymentMethodTotal(
    val method: PaymentMethod,
    val amountCents: Long,
)

data class IncomePayment(
    val id: UUID,
    val reservationId: UUID,
    val customerName: String,
    val serviceName: String,
    val amountCents: Long,
    val currency: String,
    val method: PaymentMethod,
    val paidAt: OffsetDateTime,
)

data class FinancialTrendPoint(
    val periodStart: LocalDate,
    val revenueCents: Long,
    val expenseCents: Long,
) {
    val profitCents: Long = revenueCents - expenseCents
}

data class FinancialReport(
    val from: OffsetDateTime,
    val to: OffsetDateTime,
    val currency: String,
    val revenueCents: Long,
    val expenseCents: Long,
    val operatingProfitCents: Long,
    val outstandingBalanceCents: Long,
    val completedAppointments: Int,
    val noShows: Int,
    val averageCompletedValueCents: Long,
    val granularity: TrendGranularity,
    val revenueByPaymentMethod: List<PaymentMethodTotal>,
    val trend: List<FinancialTrendPoint>,
)

class ExpenseNotFoundException(id: UUID) : RuntimeException("Expense $id was not found")

class InvalidFinanceRequestException(message: String) : RuntimeException(message)
